PropertyGraph: don't duplicate type inference logic

scober opened the pull request:
Currently we have a special case in one PropertyGraph::Make() variant
that does its own type inference. Change it to use
PropertyGraph::ConstructEntityTypeIDs() instead.

ddn0 commented:
@swreinehr can you take a look at the commit "Add unit test for querying" and let me know if this is the kind of interface you were expecting?

ddn0 commented:
cc @witchel @tylershunt for the Distribution changes.

tylershunt reviewed (commented):
Yes! I think making `Distribution` the result of both loading and partitioning makes a lot of sense. CC @scober as well since he's working in this area.

scober commented:
This sounds great. I am working on pulling some of the partitioning logic out of `Distribution` and into the partitioner, so I love anything that decouples `Distribution` from partitioning.

roshandathathri commented on libquery/src/AttributedGraph.cpp:
```diff
@@ -531,14 +529,12 @@ galois::graphs::AttributedGraph::MatchNodeDegree(
-    galois::graphs::DistGraph<std::tuple<>, std::tuple<>>&& dist_graph)
-    : dist_(dist_graph.ReleaseDistribution()) {
-  const galois::graphs::PropertyFileGraph& pfg = dist_->GetPropertyGraph();
+    galois::graphs::Distribution dist)
```
Why not take a move parameter?

ddn0 replied:
There is certainly a lot of differing opinions on what to do in this case.

- https://stackoverflow.com/questions/37935393/pass-by-value-vs-pass-by-rvalue-reference
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#fcall-parameter-passing
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f18-for-will-move-from-parameters-pass-by-x-and-stdmove-the-parameter

My motivation so far has been that users of `T` like `AttributedGraph` should not be concerned with how `T` is implemented and if the caller of `AttributedGraph` wants to give a Distribution to AttributedGraph versus making a copy (in some future world where Distributions might be copyable), this decision can be made by the caller and `AttributedGraph` itself shouldn't overly restrict things. (In absence of a strong need otherwise.)

Thinking about it more, I think just taking `T&&` for move-only types might be better because there is less chance of undefined behavior with `T&&` and perhaps it is fine for people to think of move-only as part of the fundamental nature of a class rather than an optimization. 

In this regime, we would reserve pass by value for two common exceptions:
- unique_ptr
- and function calls that take more than one class that is both copyable and efficiently movable (e.g., string) in order to avoid creating the cross-product of specializations for moving in arguments.

Thoughts?

roshandathathri reviewed (approved):
LGTM

witchel commented on libgluon/include/galois/graphs/DistGraph.h:
```diff
@@ -315,41 +313,52 @@ PartitionGraph(const DistOptions& opt) {
   }
 
+  dist.AddCommandLine(opt.command_line);
+
```
Digging this PR in general.  I find it ironic that the Distribution does not have a reference to DistOptions, but I can live with it.

ddn0 replied:
I originally named DistOptions as a reference to DistGraph. This was probably even before the extraction of Distribution from DistGraph. I think PartitionOptions might be a better name for DistOptions now. I can rename it after this PR lands if that would be an improvement.

witchel replied:
Haha.  DistGraph/DistOptions does make sense!

I don't think renaming it is worthwhile, but I did consider giving Distribution a reference to DistOptions.  But I thought people might find that profligate.  I guess there is an attempt to keep DistGraph and DistOptions at the highest, user-facing level?

ddn0 replied:
> I guess there is an attempt to keep DistGraph and DistOptions at the highest, user-facing level?

Yup. Though from a lineage perspective, I can see why passing more of the high-level data down would be desirable (e.g., command line, filename, even partitioning policy parameters like edge/node weight and partitioning rounds).

witchel commented on libgluon/include/galois/graphs/Distribution.h:
```diff
@@ -82,37 +82,56 @@ class GALOIS_EXPORT Distribution {
+  /// being fully initialized.
+  static Result<Distribution> Make(
+      const std::string& rdg_name, unsigned host, unsigned num_hosts);
+
```
I treat Distribution and Partitioner like the attic of a crazy uncle.  Interesting place, but dangerous.

I get the idea that I have one graph, I partition it, then I have many graph pieces.  What even is a Distribution?  And if I create one from an RDG, what is missing and how does the partitioner give it to me?  Where me is an RDG? a propertygraph?  a distgraph?

ddn0 replied:
I think one good outcome would be if Distribution eventually merges into RDG. There are some small things that hinder this (like the tight API coupling between the Partitioner and Distribution), but the biggest obstacle IMHO is how to structure partitioning and repartitioning (e.g., loading one partitioning or an implicit 1D partitioning and generating another one) as RDG operations.

It might be the case that PropertyFileGraph should also be merged into an RDG.

I think there will still be a need for DistGraph or PropertyGraph  but ideally not both for cases where you know statically what the types of properties are.

FWIW, I think ingest and incremental updates will force us to make these abstractions cleaner.

witchel reviewed (approved):
Anything that helps disentangle Distribution and Partitioner is welcome.

Commit:
PropertyGraph: don't duplicate type inference logic

Currently we have a special case in one PropertyGraph::Make() variant
that does its own type inference. Change it to use
PropertyGraph::ConstructEntityTypeIDs() instead.

amberhassaan reviewed (approved):
Good change.

e-mcginnis reviewed (approved):
lgtm!
